avail-server/src/main/kotlin/com/avail/server/io/files/FileActions.kt

```diff
@@ -0,0 +1,347 @@
+
+	override fun execute(file: AvailServerFile, timestamp: Long): TracedAction =
+		tracedAction
+
+	override val type: FileActionType = FileActionType.NO_ACTION
+}
+
+
+/**
+ * `UndoAction` is a [FileAction] that [executes][execute] the
+ * [inverse][TracedAction.reverseAction] for a `FileAction` that is traced as a
+ * [TracedAction].
+ *
+ * @author Richard Arriaga &lt;rich@availlang.org&gt;
+ */
+internal object UndoAction: FileAction
+{
+	override fun execute(file: AvailServerFile, timestamp: Long): TracedAction
+	{
+		file.serverFileWrapper.undo()
+		return NoAction.tracedAction
+	}
+
+	override val type: FileActionType = FileActionType.UNDO
+}
+
+/**
+ * `RedoAction` is a [FileAction] that [re-executes][execute] a recently
+ * [undone][UndoAction] [traced][TracedAction] `FileAction`.
+ *
+ * @author Richard Arriaga &lt;rich@availlang.org&gt;
+ */
+internal object RedoAction: FileAction
+{
+	override fun execute(file: AvailServerFile, timestamp: Long): TracedAction
+	{
+		file.serverFileWrapper.redo()
+		return NoAction.tracedAction
+	}
+
+	override val type: FileActionType = FileActionType.REDO
+}
+
+/**
+ * `SaveAction` is a [FileAction] that saves an [AvailServerFile] to disk.
+ *
+ * @author Richard Arriaga &lt;rich@availlang.org&gt;
+ *
+ * @property failureHandler
+ *   A function that accepts a [ServerErrorCode] that describes the nature
+ *   of the failure and an optional [Throwable]. TODO refine error handling
+ *
+ * @constructor
+ * Construct a [SaveAction].
+ *
+ * @param failureHandler
+ *   A function that accepts a [ServerErrorCode] that describes the nature
+ *   of the failure and an optional [Throwable]. TODO refine error handling
+ */
+internal class SaveAction constructor(
+	private val failureHandler: (ServerErrorCode, Throwable?) -> Unit): FileAction
+{
+	override fun execute(file: AvailServerFile, timestamp: Long): TracedAction
+	{
+		file.save(failureHandler)
+		return NoAction.tracedAction
+	}
+
+	override val type: FileActionType = FileActionType.SAVE
+}
+
+/**
+ * A `TracedAction` records a [FileAction] that was performed on a file and
+ * the `FileAction`s required to undo the initial `FileAction`.
+ *
+ * @author Richard Arriaga &lt;rich@availlang.org&gt;
+ *
+ * @property timestamp
+ *   The time when this [FileAction] request was received.
+ * @property forwardAction
+ *   The originally requested [FileAction] that was made to a file.
+ * @property reverseAction
+ *   The [FileAction] that reverses the [forwardAction].
+ *
+ * @constructor
+ * Construct a [TracedAction].
+ *
+ * @param timestamp
+ *   The time when this [FileAction] request was performed.
+ * @property forwardAction
+ *   The originally requested [FileAction] that was made to a file.
+ * @property reverseAction
+ *   The [FileAction] that reverses the `forwardAction`.
+ */
+internal class TracedAction constructor(
+	private val timestamp: Long,
+	private val forwardAction: FileAction,
+	private val reverseAction: FileAction)
```

Comment:
Per our conversation, this will have to be tagged with a User so we can have the interleaved TracedAction stack with each user maintaining their own pointers in the stack. This will include lowest possible location a user can undo to relative to a potential concurrent change made by another user that "breaks" all undos for you below that point. We'll add functionality to compare changes to ensure where this all happens. This will also include "external changes" tag for changes to a file outside the IDE. The "external changes" tag can be used for identifying changes by other users for security purposes when a user isn't allowed to know about the changes.